### PR TITLE
Added a support for gzipped fie in planetaryimage.

### PR DESCRIPTION
**Pull request by bvnayak**
refs #42

### CODE REVIEW

File: planetaryimage/image.py [hunk 1]
```diff
@@ -274,7 +279,7 @@ def _parse_data(self, stream):
             data_stream.close()
 
     def _parse_band_sequential_data(self, stream):
-        data = numpy.fromfile(stream, self.dtype, self.size)
+        data = numpy.fromstring(stream.read(self.size*4), self.dtype)
```
Comment (bvnayak) [inline planetaryimage/image.py]: As, numpy.fromfile supports `open_file_type` and not the `gzip_open_fie_type`, numpy.fromstring used.

File: tests/test_pds3file.py [hunk 2]
```diff
@@ -49,6 +72,14 @@ def test_pds3_1band_disk_format(pattern_data):
...
     assert_almost_equal(image.data[0], pattern_data)
 
+def test_gz_pds3_1band_disk_format(pattern_data):
+    image = PDS3Image(gzip.open(gzipped_filename, 'rb'), memory_layout='DISK')
+
+    assert image.data.shape == (1, 10, 10)
+    assert image.data.size == 100
+
+    assert_almost_equal(image.data[0], pattern_data)
+
```
Comment (bvnayak) [inline tests/test_pds3file.py]: screen shot 2015-09-22 at 12 43 38 pm

File: planetaryimage/image.py (hunk 1 above)
Comment (wtolson) [inline planetaryimage/image.py]: This going to cause an extra copy and double the memory for all cases which can be significant for large images. I suggest we default to `numpy.fromfile` and fallback to `numpy.fromstring`, something like:

    try:
        data = numpy.fromfile(stream, self.dtype, self.size)
    except IOError:
        data = numpy.fromstring(stream.read(self.size*4), self.dtype)

**Comment by wtolson** [comment]
It might be nice to implement an interface for handling compression similar to `pandas.read_csv()`. It takes the following `compression` argument:

> compression: decompress 'gzip' and 'bz2' formats on the fly. Set to 'infer' (the default) to guess a format based on the file extension.

This will allow additional compression formats like bz2 from the stdlib as well as allowing the user to override and disable the compression.

File: planetaryimage/image.py [hunk 3]
```diff
@@ -26,10 +28,17 @@ def open(cls, filename):
         filename : string
             name of file to read as an image file
         """
-        with open(filename, 'rb') as fp:
-            return cls(fp, filename)
+        if filename.endswith('.gz'):
+            with gzip.open(filename, 'rb') as fp:
+                return cls(fp, filename, compression='gz')
```
Comment (bvnayak) [inline planetaryimage/image.py]: Added keyword arg `compression` which is `none` by default.

File: planetaryimage/image.py [hunk 4]
```diff
@@ -274,7 +285,10 @@ def _parse_data(self, stream):
             data_stream.close()
 
     def _parse_band_sequential_data(self, stream):
-        data = numpy.fromfile(stream, self.dtype, self.size)
+        if self.compression == 'none':
```
Comment (bvnayak) [inline planetaryimage/image.py]: if `compression` is `none` then use `numpy.fromfile` else `numpy.fromstring`.
Reason:
    numpy.fromfile returns '[]' for gzip or bz2 open file and won't throw any error.

File: tests/test_pds3file.py [hunk 5]
```diff
@@ -2,6 +2,8 @@
 import pytest
 import os
 import numpy
+import gzip
```
Comment (bvnayak) [inline tests/test_pds3file.py]: Travis CI fails for Python 2.6 (for gzip and bz2) and for Python 2.7 (bz2).
Working on resolving this issue with Python Versions 2.6 and 2.7.

**Comment by godber** [comment]
@wtolson your suggestion of catching an IOError was a good idea but an error was not being raised, `fromfile()` was simply returning a null array.  So we opted for the explicit route.

Now we just have to figure out how to deal with the python 2.6 and 2.7 issues.

**Comment by wtolson** [comment]
Interesting, what behaviour are you seeing? On python2.7/numpy1.9.2 I'm seeing this:

    In [1]: import gzip

    In [2]: import numpy

    In [3]: with gzip.open('test.txt.gz', 'rb') as fp:
        data = numpy.fromfile(fp, int, 10)
       ...:     
    ---------------------------------------------------------------------------
    IOError                                   Traceback (most recent call last)
    <ipython-input-3-74cf4401a888> in <module>()
          1 with gzip.open('test.txt.gz', 'rb') as fp:
    ----> 2     data = numpy.fromfile(fp, int, 10)
          3 

    IOError: first argument must be an open file

    In [4]: print numpy.version.full_version
    1.9.2

**Comment by godber** [comment]
I think Bhavin is workin in python 3 by default.  He's in class at the moment.

File: planetaryimage/image.py [hunk 6]
```diff
@@ -26,10 +29,25 @@ def open(cls, filename):
...
         """
-        with open(filename, 'rb') as fp:
-            return cls(fp, filename)
+        if filename.endswith('.gz'):
+            try:
+                with gzip.open(filename, 'rb') as fp:
+                    return cls(fp, filename, compression='gz')
+            except AttributeError:
+                raise IOError('GZIP compression not supported until Python 2.6.')
+                return
```
Comment (godber) [inline planetaryimage/image.py]: returns not needed after the raise.

File: planetaryimage/image.py [hunk 7]
```diff
@@ -26,10 +29,25 @@ def open(cls, filename):
...
             name of file to read as an image file
         """
-        with open(filename, 'rb') as fp:
-            return cls(fp, filename)
+        if filename.endswith('.gz'):
+            try:
+                with gzip.open(filename, 'rb') as fp:
+                    return cls(fp, filename, compression='gz')
+            except AttributeError:
+                raise IOError('GZIP compression not supported until Python 2.6.')
```
Comment (godber) [inline planetaryimage/image.py]: I thought gzip was not available until 2.7 and bz2 was not available until 3.3?

File: planetaryimage/image.py [hunk 8]
```diff
@@ -26,10 +29,25 @@ def open(cls, filename):
...
+                    return cls(fp, filename, compression='bz2')
+            except AttributeError:
+                raise IOError('BZ2 compression not supported until Python 2.7.')
+                return
+        else:
+            with open(filename, 'rb') as fp:
+                return cls(fp, filename)
 
-    def __init__(self, stream, filename=None, memory_layout='DISK'):
+    def __init__(self, stream, filename=None, compression='none', memory_layout='DISK'):
```
Comment (godber) [inline planetaryimage/image.py]: Using `compression=None` would be more pythonic.

File: planetaryimage/image.py [hunk 9]
```diff
@@ -26,10 +29,25 @@ def open(cls, filename):
...
+        if filename.endswith('.gz'):
+            try:
+                with gzip.open(filename, 'rb') as fp:
+                    return cls(fp, filename, compression='gz')
+            except AttributeError:
+                raise IOError('GZIP compression not supported until Python 2.6.')
+                return
+        elif filename.endswith('.bz2'):
+            try:
+                with bz2.open(filename, 'rb') as fp:
```
Comment (wtolson) [inline planetaryimage/image.py]: We can use [`bz2.BZ2File`](https://docs.python.org/2.6/library/bz2.html#bz2.BZ2File) and [`gzip.GzipFile`](https://docs.python.org/2.6/library/gzip.html#gzip.GzipFile) which will give us identical support back to python2.6. I'm not sure we should support python2.5. Thoughts @godber?

Comment (godber) [inline planetaryimage/image.py]: Will they work inside the `with` context managers?  I think that was the problem in at least one of the cases.

Comment (godber) [inline planetaryimage/image.py]: Oh and lets not bother with 2.5, 2.6 is still a stretch but I think its still default on RHEL6, which makes it relevant.

File: planetaryimage/image.py (hunk 7 above)
Comment (bvnayak) [inline planetaryimage/image.py]: gzip with context support is available from 2.7 and bz2 from 3.3. So, gzip won't be available till 2.6 and bz2 till 2.7.

**Comment by bvnayak** [comment]
[GZIP: Changed in version 2.7: Support for the with statement was added.](https://docs.python.org/2/library/gzip.html)

[BZ2.open is new in Version 3.3](https://docs.python.org/3.3/library/bz2.html)

File: planetaryimage/image.py (hunk 7 above)
Comment (bvnayak) [inline planetaryimage/image.py]: It was a grammatical mistake. Corrected in the latest commit.

File: planetaryimage/image.py (hunk 9 above)
Comment (wtolson) [inline planetaryimage/image.py]: Python 2.6 has really bad support for context managers in the standard lib. This is the type of code it expects:

    stream = bz2.BZ2File(filename, 'rb')

    try:
        return cls(stream, filename)
    finally:
        stream.close()

Comment (godber) [inline planetaryimage/image.py]: Oh, I never thought of using `finally`.  That seems like a good approach.  I thought we would have to track down all of the right places to close the stream and that seemed like a bad idea.

**Pull request merged**